### ISSUE TITLE
Consider array for all Node structs and other possible optimizations

### ISSUE THREAD

**Issue by djhoese** [storpipfugl/pykdtree#92]
In #81, when talking about serializing the KDTree object I pointed out that the Tree/Node structure internal to pykdtree is structured as a series of `struct` objects along with an "ancillary" `pidx` arrray (size of points) and a `bbox` (size of dimensions * 2).

I've been contemplating whether it would be smarter to declare the tree nodes as a large array and fill it in as we build the tree. This has a couple advantages as well as makes it easier to do some other refactorings. The Tree/Node array could be structured in a depth first or breadth first manner depending on what is more performant. Given the way the algorithm works I think it would be smarter to structure this faster building rather than faster querying, but I'm not sure which is which.

If implemented, I see this changing a few things:

1. Serialization should be much easier (see #81) since the tree/nodes are no longer custom C-level structs, but could just be a buffer of bytes. It would be large, but no larger than the existing total amount of memory.
2. I *think* overall memory usage should be smaller since we don't need have pointers to left/right nodes. Leaf-ness (whether a node is a leaf or not) is just based on how many elements are left in the array.
3. Tree building should be faster because we no longer have to do a `malloc` to create each Node. I would assume this should allow the compiler to optimize things even further and not wait for memory to be allocated.
4. Querying *might* also get sped up, but since the point of the tree structure is to visit as few nodes as possible it wouldn't be the primary goal.
5. This (I think) would make it easier to restructure the recursive functions into non-recursive/`for` loop set of functions. This should (again, I think) allow more optimizations from the compiler.

Thoughts? I don't have time to implement this, but I'd be curious what kind of difference it makes.

**Comment by Dementiy** [comment]
Hello. Is there any progress on this issue?

**Comment by djhoese** [comment]
No. Would you like to take a look at implementing this @Dementiy? I made this issue as a "when I magically have time, I bet this could be fun", but no available time from me for something this big.

**Comment by Dementiy** [comment]
I'm not good at C, but I'll definitely take a look at this in a couple of weeks.

**Comment by djhoese** [comment]
I took a look at this a bit this weekend just to try to understand how the C code here actually works. The situation isn't quite as bad as I had thought, but I do think there is still room for improvement. The main thing is that the `pidx` or "permutation index", which holds the indexes of the source points in kd-tree order, is one contiguous block of memory allocated at the start of tree creation. The individual Nodes are still created one at a time which could still be done.

Although much harder to implement, there could also be some thought put into memory efficiency of not having the final `pidx` at all and instead making it part of the leaf nodes themselves. As it is right now we essentially have a series of linked structs spread across memory:

```
tree -> root node -> node -> ... -> node -> leaf node -> index into pidx -> index into source data
```

So indexing into a contiguous block of memory in `pidx` and then into a contiguous block of memory into the starting coordinates (points) and then comparing that against one of the provided query points. So that's 4 different types of memory being accessed if I have this right: tree nodes (non-contiguous), pidx, source points (`pa` in code), and query points. If there was a way to have the tree nodes in contiguous memory *and* have the source points next to them, then I would think the kernel/system could efficiently pull useful blocks of memory. There could also be thought into whether left or right node children should be next to the parent. This would get memory access down to 3 different types instead of 4 and hopefully overall better caching.

Anyway...the above proposed memory structure could be very complicated and could waste a lot of memory if the tree is not well balanced as you'd have to pre-allocate a known number of locations for leaf data points whether they were all used or not. It would essentially have to be a random blob of bytes that get cast at runtime into node structs and an index type (ex. `uint32` or `uint64` in the current code).

Lastly, without thinking about it too much, I wonder if it would be worth it to try to put nodes, pidx, and the source points into one large structure. Nah...that gets real ugly as the memory after your leaf nodes would be the source index (which needs to be returned from a query) followed by `number of leaf elements * number of dimensions`.

Ok, done braindumping.
